TODOs implemented. Seems to bo working.

Deomitri93 opened the pull request.

Commit:
Commit 21.11.2020 (TODOs implemented. Seems to bo working.)

petrelevich commented on hw07-gradle/src/main/java/ru/otus/processor/homework/LoggerProcessor2SecondsException.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public Message process(Message message) {
+        if (LocalDateTime.now().getSecond() % 2 == 0) {
```
при таком варианте почти невозможно выполнить требование
```
(сделайте тест с гарантированным результатом)
```
Лучше сделать интерфейс - dataProvider.
В тестах имплементация будет - нужная константа
в боевом применении - настоящее время.

Deomitri93 replied:
Добавил DataProvider, который в боевом применении использует реализацию с LocalTime.now().getSecond().
В тестах же использую две реализации: одна всегда возвращает четную секунду, другая - нечетную.

petrelevich commented on hw07-gradle/src/main/java/ru/otus/listener/homework/ListenerMessageHistory.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+public class ListenerMessageHistory implements Listener {
+    private List<String> msgHistory = new ArrayList<>();
```
не очень универсальное решение хранить строку.
Если потребуется дальнейшая обработка, придется парсить.
Лучше сделать класс - Record, в котором будут старое и новое значения.
Экземпляр этого класса можно будет при необходимости в разных формтах представлять.

Deomitri93 replied:
Добавил inner класс MessageHistoryRecord.

Deomitri93 committed:
24.11.2020
  Record class for message history storage added;
  DataProvider for getting current second added;
  Test for throwing exception processor added;

petrelevich commented on hw07-gradle/src/main/java/ru/otus/listener/homework/ListenerMessageHistory.java:
```diff
@@ -0,0 +1,42 @@
+
+        public MessageHistoryRecord(Message oldMessage, Message newMessage, LocalDate dateMessage) {
+            this.oldMessage = oldMessage;
+            this.newMessage = newMessage;
```
тут ты ссылку на объект сохраняете, а объект не простой.
private final ObjectForMessage field13;
при измении этого поля сломается история.
Надо делать глубокую копию.

Deomitri93 replied:
Поддержал в классах Message и ObjectForMessage интерфейс Serializable и реализовал deep copy с помощью сериализации полей MessageHistoryRecord.

petrelevich replied:
Сериализация для этого не предназначена.
Лучше скопировать нужные поля "руками".

Deomitri93 replied:
Сделал с помощью clone().

Deomitri93 replied:
Переделал с помощью конструктора копирования

Deomitri93 committed:
30.11.2020 (Deep copy via serialization implemented)

Deomitri93 committed:
01.12.2020 (Cloning implemented)

Deomitri93 committed:
02.12.2020 (Copy constructor)
